File: Crio/crio_programming_interview_problems-master/RobotInAGrid/RobotInAGrid.java
import java.util.*;

public class RobotInAGrid {

    //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int visit[][];
    ArrayList<Integer>path;
    Vector<String> pathString;
    boolean isPath;
    boolean isValid(int x , int y , int n , int m , int grid[][]) {
        if(x < 0 || x >= n || y < 0 || y >= m) {
            return false;
        }
        if(grid[x][y] == 1 || visit[x][y] == 1) {
            return false;
        }
        return true;
    }
    void dfs(int x , int y , int grid[][] , int n , int m) {
        visit[x][y] = 1;
        path.add(x);
        path.add(y);
        if(x == n-1 && y == m-1) {
            for(int i = 0 ; i < path.size() ; i+=2) {
                pathString.add(Integer.toString(path.get(i)+1)+" "+Integer.toString(path.get(i+1)+1));
            }
            isPath = true;
            return;
        } 
        if(isValid(x+1, y, n, m, grid)) {
            dfs(x+1, y, grid , n , m);
        }
        if(isValid(x, y+1, n, m, grid)) {
            dfs(x, y+1, grid , n , m);
        }
        path.remove(path.size()-1);
        path.remove(path.size()-1);

    }
    //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    Vector<String> robotInAGrid(int grid[][] , int n , int m) {

        //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        isPath = false;
        visit = new int[n][m];
        path = new ArrayList<>();
        pathString = new Vector<String>();
        for(int i = 0 ; i < n ; i++) {
            for(int j = 0 ;j < m ; j++) {
                visit[i][j] = 0;
            }
        }
        if(grid[0][0] == 1)
        {  
            Vector<String> st = new Vector<String>();
            st.add("Not Possible");
            return st;
        }
        dfs(0, 0, grid, n, m);
        if(isPath) {
            return pathString;
        } else {
            Vector<String> notposs = new Vector<String>();
            notposs.add("Not Possible");
            return notposs;
        }
        //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        int grid[][] = new int[n][m];
        for(int i = 0 ; i < n ; i++) {
            for(int j = 0 ; j < m ; j++) {
                grid[i][j] = scanner.nextInt();
            }
        }
        scanner.close();
        RobotInAGrid answer = new RobotInAGrid();
        Vector<String> result = answer.robotInAGrid(grid , n , m);
        for(String elem : result)
        {
            System.out.println(elem);
        }
    }
}

File: Crio/crio_programming_interview_problems-master/RobotInAGrid/RobotInAGrid.py
import sys

class RobotInAGrid:
    def __init__(self, rows, columns):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        self.visited = [[0 for i in range(columns)] for j in range(rows)] 
        self.path = list()
        self.isPath = None
        self.pathString = list()

    def isValid(self,x, y, n, m, grid):
        if x<0 or x>=n or y<0 or y>=m:
            return False
        if grid[x][y] == 1 or self.visited[x][y] == 1:
            return False
        return True


    def dfs(self,x,y,grid,n,m):
        self.visited[x][y]=1
        self.path.extend([x,y])
        if x== n-1 and y== m-1:
            for i in range(0,len(self.path),2):
                self.pathString.append(' '.join([str(j+1) for j in self.path[i:i+2]]))
            self.isPath = True
            return 
        if self.isValid(x+1,y,n,m,grid):
            self.dfs(x+1,y,grid,n,m)
        if self.isValid(x,y+1,n,m,grid):
            self.dfs(x,y+1,grid,n,m)
        self.path = self.path[:-2]
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    def robotInAGrid(self, grid, n, m):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        self.isPath = False
        self.path = list()
        self.pathString = list()
        if grid[0][0] == 1:
            return ["Not Possible"]
        self.dfs(0,0,grid,n,m)
        if self.isPath:
            return self.pathString
        else:
            return ["Not Possible"] 

# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    

def main():
    rows, columns = map(int,input().split())
    grid=list()
    for _ in range(rows):
        grid.append(list(map(int,input().split())))
    obj =  RobotInAGrid(rows,columns) # Any initialization if needed
    result = obj.robotInAGrid(grid,rows,columns)
    for elem in result:
        print(elem)


if __name__ == "__main__":
    sys.setrecursionlimit(10 * 1000)
    main()
